torch/ao/pruning/_experimental/pruner/README.md

```diff
@@ -1,93 +1,263 @@
+```python
+# need to enable mask update
+pruner.enable_mask_update = True
+pruner.step()
+```
 
+After this step is called, the masks will be updated to prune the corresponding weights. You can expect the ouput of the model after this step to be the same as the pruned model.
 
+However, there will be no performance / memory benefits for the model, since the pruned parameters are still present but just zeroed out. In order to achieve performance gains, we need to resize the weight tensors, by calling `prune`.
+```python
+# to prune the module
+pruned_model = pruner.prune()
 ```
-pruner = ImplementedPruner(defaults=None, also_prune_bias=True)
-pruner.prepare(model, config)
-```
 
+Prune will symbolically trace the model in order to get a graph, and then search that graph for particular patterns (op sequences)
+
+Once prune identifies an op sequence, it is able to resize the weights by calling the pruning function for a paticular block. This is defined by a dictionary that maps patterns (tuple of nn modules / functions) to a particular pruning function. You can extend this pattern dictionary to add support for pruning different modules / patterns.
 
-`step`: applies `update_mask` logic (i.e. prunes the weight matrix)
+For this particular example, we would expect to match the `(nn.Linear, nn.ReLU, nn.Linear)` pattern in our default pattern dict.
 
+Once this is done for all matches, `prune` returns your pruned model:
 
 ```
-pruner.step()
+# original model
+seq.0.weight        | 500, 700        |    350000
+seq.0.bias          | 500             |       500
+seq.2.weight        | 800, 500        |    400000
+seq.4.weight        | 600, 800        |    480000
+seq.4.bias          | 600             |       600
+linear.weight       | 4, 600          |      2400
+=== Total Number of Parameters: 1233500 ===
+```
+```
+# pruned model
+seq.0.weight        | 250, 700        |    175000
+seq.0.bias          | 250             |       250
+seq.2.weight        | 400, 250        |    100000
+seq.4.weight        | 300, 400        |    120000
+seq.4.bias          | 300             |       300
+linear.weight       | 2, 300          |       600
+=== Total Number of Parameters: 396150 ===
 ```
 
+Note that since we prune 50% of the output rows, this also corresponds to pruning 50% of the input columns for the subsequent layer.
+This is why our total number of parameters is roughly (1-0.5)* (1-0.5) = 0.25 of the original number of parameters.
 
-`squash_mask`: applies the parametrization one last time to the weight matrix, and then removes the pruning parametrization from the model
-
+## Writing Custom Patterns and Pruning Functions for Structured Pruning
+What should you do if the layers you are trying to prune don't match the existing patterns?
 
-```
-pruner.squash_mask()
-```
+If you're working with linear/conv2d layers, it's very probable that you just need to add an entry to the pattern dict mapping your pattern to an existing prune_function.
```

Comment:
unrelated to the documentation, this API seems a bit painful for the average user, if all the user is doing most of the time is indicating that in their (conv-x-conv) pattern that the x op doesn't need to be pruned, the API should probably be able to handle that more elegantly than requiring a user to define a new function that essentially just calls the normal conv2d pruner. If thats not true, i.e. the average user actually needs to carefully handle the x op pruning in most anticipated cases, I think this guide should probably have an example going through that

Reply:
I think the goal is that for most of these patterns, they are already present in the default pruning pattern dict. So for most users this section is unnecessary, and all they have to interact with is the code in the section above.

I think I might split this section and the Implementing a pruner section out, these are more for power users and probably don't need to be in the tutorial.